iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,325 @@
+                                }
+                                None => {
+                                    debug!("endpoint closed");
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    // close all connections and wait until they have all grace
+                    // fully closed.
+                    endpoint.close(QUIC_ADDR_DISC_CLOSE_CODE, QUIC_ADDR_DISC_CLOSE_REASON);
+                    endpoint.wait_idle().await;
+
+                    // all tasks should be closed, since the endpoint has shutdown
+                    // all connections, but await to ensure they are finished.
+                    set.abort_all();
+                    while !set.is_empty() {
+                        _ = set.join_next().await;
+                    }
+
+                    debug!("quic endpoint has been shutdown.");
+                }
+                .instrument(info_span!("quic-endpoint")),
+            );
+            Ok(Self {
+                bind_addr,
+                cancel,
+                handle: AbortOnDropHandle::new(task),
+            })
+        }
+
+        /// Closes the underlying QUIC endpoint and the tasks running the
+        /// QUIC connections.
+        pub async fn shutdown(mut self) -> Result<()> {
+            self.cancel.cancel();
+            if !self.task_handle().is_finished() {
+                self.task_handle().await?
+            }
+            Ok(())
+        }
+    }
+
+    /// A handle for the Server side of QUIC address discovery.
+    ///
+    /// This does not allow access to the task but can communicate with it.
+    #[derive(Debug, Clone)]
+    pub struct ServerHandle {
+        cancel_token: CancellationToken,
+    }
+
+    impl ServerHandle {
+        /// Gracefully shut down the quic endpoint.
+        pub fn shutdown(&self) {
+            self.cancel_token.cancel()
+        }
+    }
+
+    /// Handle the connection from the client.
+    ///
+    /// Any errors that happen during this connection do not need to be handled,
+    /// and will be logged at the debug level in this function.
+    async fn handle_connection(incoming: quinn::Incoming) -> Result<()> {
+        let connection = match incoming.await {
+            Ok(conn) => conn,
+            Err(e) => {
+                debug!("error accepting incoming connection: {e:#?}");
+                return Err(e.into());
+            }
+        };
+        debug!("established");
+        // wait for the client to close the connection
+        let connection_err = connection.closed().await;
+        match connection_err {
+            quinn::ConnectionError::ApplicationClosed(ApplicationClose { error_code, .. })
+                if error_code == QUIC_ADDR_DISC_CLOSE_CODE =>
+            {
+                Ok(())
+            }
+            _ => {
+                debug!("error closing connection {connection_err:#?}",);
```

Comment:
"connection closed with error"?

This side is not doing the active closing I guess.

Reply:
tru